src/stacked_borrows.rs

```diff
@@ -325,32 +291,60 @@ impl<'tcx> Stack {
         provoking_access: Option<(SbTagExtra, AllocRange, Size, AccessKind)>, // just for debug printing and error messages
         global: &GlobalStateInner,
         alloc_history: &mut AllocHistory,
+        threads: &ThreadManager<'_, 'tcx>,
     ) -> InterpResult<'tcx> {
-        if global.tracked_pointer_tags.contains(&item.tag) {
+        if global.tracked_pointer_tags.contains(&item.tag()) {
             register_diagnostic(NonHaltingDiagnostic::PoppedPointerTag(
                 *item,
                 provoking_access.map(|(tag, _alloc_range, _size, access)| (tag, access)),
             ));
         }
 
-        if let Some(call) = item.protector {
-            if global.is_active(call) {
-                if let Some((tag, _alloc_range, _offset, _access)) = provoking_access {
-                    Err(err_sb_ub(
-                        format!(
-                            "not granting access to tag {:?} because incompatible item is protected: {:?}",
-                            tag, item
-                        ),
-                        None,
-                        tag.and_then(|tag| alloc_history.get_logs_relevant_to(tag, Some(item.tag))),
-                    ))?
-                } else {
-                    Err(err_sb_ub(
-                        format!("deallocating while item is protected: {:?}", item),
-                        None,
-                        None,
-                    ))?
-                }
+        if !item.protected() {
+            return Ok(());
+        }
+
+        // We store tags twice, once in global.protected_tags and once in each call frame.
+        // We do this because consulting a single global set in this function is faster
+        // than attempting to search all call frames in the program for the `FrameExtra`
+        // (if any) which is protecting the popped tag.
+        //
+        // This duplication trades off making `end_call` slower to make this function faster. This
+        // trade-off is profitable in practice for a combination of two reasons.
+        // 1. A single protected tag set can (and does in some programs) protect thousands of `Item`s.
```

Comment:
```suggestion
        // 1. A single protected tag can (and does in some programs) protect thousands of `Item`s.
```
At least I think that's what you mean?

Reply:
Yup.